ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/MaximumLagFilter.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2020 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.rest.server;
+
+import static java.util.Objects.requireNonNull;
+
+import com.google.common.collect.ImmutableMap;
+import io.confluent.ksql.execution.streams.RoutingFilter;
+import io.confluent.ksql.execution.streams.RoutingOptions;
+import io.confluent.ksql.rest.entity.LagInfoEntity;
+import io.confluent.ksql.rest.entity.QueryStateStoreId;
+import io.confluent.ksql.util.KsqlHostInfo;
+import java.util.List;
+import java.util.Optional;
+import java.util.OptionalLong;
+import java.util.function.Function;
+
+/**
+ * A RoutingFilter that filters hosts based upon changelog processing lag.
+ */
+public final class MaximumLagFilter implements RoutingFilter {
+
+  private final ImmutableMap<KsqlHostInfo, Optional<LagInfoEntity>> lagByHost;
+  private final RoutingOptions routingOptions;
+  private final OptionalLong maxEndOffset;
+
+  private MaximumLagFilter(
+      final RoutingOptions routingOptions,
+      final ImmutableMap<KsqlHostInfo, Optional<LagInfoEntity>> lagByHost,
+      final OptionalLong maxEndOffset
```

Comment:
can we document this parameter? it's a little unclear what it means to me (i.e. why would I prefer this over whats in the map - `maxEndOffset.orElse(hostLag.getEndOffsetPosition())`)

EDIT: after reading more of the code this makes a little more sense to me - it's the latest end offset that any of the replicas has seen (so it's independent of how caught up any of them actually are).

Reply:
Yep, you got it.

I added more documentation to this constructor as well as the filtering itself.